src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs

```diff
@@ -808,6 +808,11 @@ public RequestLineStatus TakeStartLine(SocketInput input)
 
                 ConnectionControl.CancelTimeout();
 
+                if (scan.Peek() == 0x16)
+                {
+                    throw new IOException("SSL/TLS handshake was attempted on an HTTP endpoint.");
```

Comment:
Yeah, I'll do that. I had initially reasoned that this would end up with the server attempting to send a 400 response which make no sense to a TLS client, but actually if I throw before changing `_requestProcessingStatus` (like I'm doing here) that won't happen.